### PR TITLE
INT-3235 - Add initial integration

### PR DESCRIPTION
**Pull request by eXtremeX**
# Description

Thank you for contributing to a JupiterOne integration!

Please include a summary of the change and which issue is fixed. Please also
include relevant motivation and context. List any dependencies that are required
for this change.

## Summary

image

The following entities are created:

| Resources              | Entity `_type`                     | Entity `_class` |
| ---------------------- | ---------------------------------- | --------------- |
| Account                | `hashicorp_vault_account`          | `Account`       |
| Authentication Backend | `hashicorp_vault_auth_backend`     | `Service`       |
| Cubbyhole Engine       | `hashicorp_vault_cubbyhole_engine` | `Service`       |
| Cubbyhole Secret       | `hashicorp_vault_cubbyhole_secret` | `Secret`        |
| KV1 Engine             | `hashicorp_vault_kv1_engine`       | `Service`       |
| KV1 Secret             | `hashicorp_vault_kv1_secret`       | `Secret`        |
| KV2 Engine             | `hashicorp_vault_kv2_engine`       | `Service`       |
| KV2 Secret             | `hashicorp_vault_kv2_secret`       | `Secret`        |
| User                   | `hashicorp_vault_user`             | `User`          |

### Relationships

The following relationships are created:

| Source Entity `_type`              | Relationship `_class` | Target Entity `_type`              |
| ---------------------------------- | --------------------- | ---------------------------------- |
| `hashicorp_vault_account`          | **HAS**               | `hashicorp_vault_auth_backend`     |
| `hashicorp_vault_account`          | **HAS**               | `hashicorp_vault_cubbyhole_engine` |
| `hashicorp_vault_account`          | **HAS**               | `hashicorp_vault_kv1_engine`       |
| `hashicorp_vault_account`          | **HAS**               | `hashicorp_vault_kv2_engine`       |
| `hashicorp_vault_auth_backend`     | **HAS**               | `hashicorp_vault_user`             |
| `hashicorp_vault_cubbyhole_engine` | **HAS**               | `hashicorp_vault_cubbyhole_secret` |
| `hashicorp_vault_kv1_engine`       | **HAS**               | `hashicorp_vault_kv1_secret`       |
| `hashicorp_vault_kv2_engine`       | **HAS**               | `hashicorp_vault_kv2_secret`       |

## Type of change

Please leave any irrelevant options unchecked.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [X] This change requires a documentation update

### General Development Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### Integration Development Checklist:

Please leave any irrelevant options unchecked.

- [X] I have checked for additional permissions required to call any new API
      endpoints, and have documented any additional permissions in
      `jupiterone.md`, where necessary.
- [X] My changes properly paginate the target service provider's API
- [X] My changes properly handle rate limiting of the target service provider's
      API
- [X] My new integration step is instrumented to execute in the correct order
      using `dependsOn`
- [X] I have referred to the
      [JupiterOne data model](https://github.com/JupiterOne/data-model/tree/main/src/schemas)
      to ensure that any new entities/relationships, and relevant properties,
      match the recommended model for this class of data
- [X] I have updated the `CHANGELOG.md` file to describe my changes
- [X] When changes include modifications to existing graph data ingestion, I've
      reviewed all existing managed questions referencing the entities,
      relationships, and their property names, to ensure those questions still
      function with my changes.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,3 +7,27 @@ and this project adheres to
...
 ## [Unreleased]
+
+### Added
+
+- Ingest new entities
+
+  - `hashicorp_vault_account`
+  - `hashicorp_vault_auth_backend`
+  - `hashicorp_vault_cubbyhole_engine`
+  - `hashicorp_vault_cubbyhole_secret`
```
Comment (austinkelleher) [inline CHANGELOG.md:18]: Consider updating to follow this J1QL:

```
FIND hashicorp_vault_secret WITH source="cubbyhole"
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,3 +7,27 @@ and this project adheres to
...
+  - `hashicorp_vault_kv1_engine`
```
Comment (austinkelleher) [inline CHANGELOG.md:19]: Consider the following model update:

```
FIND hashicorp_vault_engine WITH name="kv1"
```

Relationships can then be queried like this:

```
FIND hashicorp_vault_engine THAT HAS hashicorp_vault_secret
```

File: src/steps/auth-backend/index.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  instance,
+  jobState,
+}: IntegrationStepExecutionContext<IntegrationConfig>) {
+  const apiClient = createAPIClient(instance.config);
+
+  const accountEntity = (await jobState.getData(ACCOUNT_ENTITY_KEY)) as Entity;
+  const authMethods = (await apiClient.getMounts()).data.auth;
+
+  for (const [key, value] of Object.entries(authMethods)) {
+    if (!['userpass'].includes(value.type)) {
```
Comment (austinkelleher) [inline src/steps/auth-backend/index.ts:25]: We should be sure to call out the fact that only `userpass` auth method users are being ingested in our docs.

File: src/steps/auth-backend/converter.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  authName: string,
+  auth: HashiCorpVaultAuth,
+): Entity {
+  return createIntegrationEntity({
+    entityData: {
+      source: {
+        ...auth,
+        name: authName,
+      },
+      assign: {
```
Comment (austinkelleher) [inline src/steps/auth-backend/converter.ts:19]: It looks like there are only required data model properties described below. Are there other properties on the `HashiCorpVaultAuth` that we should add? Not sure whether this TypeScript type is correct or not: https://github.com/JupiterOne/graph-hashicorp-vault/pull/5/files#diff-c54113cf61ec99691748a3890bfbeb00e10efb3f0a76f03a0fd9ec49072e410aR61

File: docs/jupiterone.md [hunk 5]
```diff
@@ -96,21 +147,32 @@ https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
...
-| `acme_account`        | **HAS**               | `acme_group`          |
-| `acme_account`        | **HAS**               | `acme_user`           |
-| `acme_group`          | **HAS**               | `acme_user`           |
+| Source Entity `_type`              | Relationship `_class` | Target Entity `_type`              |
+| ---------------------------------- | --------------------- | ---------------------------------- |
+| `hashicorp_vault_account`          | **HAS**               | `hashicorp_vault_auth_backend`     |
+| `hashicorp_vault_account`          | **HAS**               | `hashicorp_vault_cubbyhole_engine` |
+| `hashicorp_vault_account`          | **HAS**               | `hashicorp_vault_kv1_engine`       |
+| `hashicorp_vault_account`          | **HAS**               | `hashicorp_vault_kv2_engine`       |
+| `hashicorp_vault_auth_backend`     | **HAS**               | `hashicorp_vault_user`             |
```
Comment (austinkelleher) [inline docs/jupiterone.md:172]: We should probably consider adding a direct relationship from the `Account` to the `User` for consistency as well.

File: docs/jupiterone.md [hunk 6]
```diff
@@ -96,21 +147,32 @@ https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
...
 The following entities are created:
 
-| Resources | Entity `_type` | Entity `_class` |
-| --------- | -------------- | --------------- |
-| Account   | `acme_account` | `Account`       |
-| User      | `acme_user`    | `User`          |
-| UserGroup | `acme_group`   | `UserGroup`     |
+| Resources              | Entity `_type`                     | Entity `_class` |
+| ---------------------- | ---------------------------------- | --------------- |
+| Account                | `hashicorp_vault_account`          | `Account`       |
```
Comment (austinkelleher) [inline docs/jupiterone.md:152]: Talking to @eXtremeX, we are thinking of simplifying the entities to the following:

```
hashicorp_vault_account | Account
hashicorp_vault_auth_backend | Service
hashicorp_vault_secret | Secret
hashicorp_vault_engine | Service
hashicorp_vault_user | User
```

File: src/steps/auth-backend/index.ts (hunk 3 above)
Comment (eXtremeX) [reply src/steps/auth-backend/index.ts:25]: The jupiterone.md has this paragraph already:

### About users

The integration can fetch all the users that exist in `userpass` authentication
method. If you'd like this, you need to make sure that `userpass` method of your
choosing (if you've got multiple) is defined in the policy assigned to the
token.

Let me know if it should be worded differently!

**Comment by eXtremeX** [comment]
Hey @austinkelleher please take another look - the changes you mentioned should be covered now!

**Review by neurovelho — APPROVED** [review]

**Pull request merged**